src/core/TChem_Driver.cpp

```diff
@@ -0,0 +1,279 @@
+  policy_type policy(exec_space_instance, 1, Kokkos::AUTO());
+
+  const ordinal_type level = 1;
+  ordinal_type per_team_extent(0);
+
+  per_team_extent = TChem::AtmosphericChemistry::getWorkSpaceSize(_kmcd_device);
+
+  const ordinal_type per_team_scratch =
+      TChem::Scratch<real_type_1d_view>::shmem_size(per_team_extent);
+  policy.set_scratch_size(level, Kokkos::PerTeam(per_team_scratch));
+
+  auto number_of_equations = problem_type::getNumberOfTimeODEs(_kmcd_host);
+  real_type_2d_view tol_time("tol time", number_of_equations, 2);
+  real_type_1d_view tol_newton("tol newton", 2);
+  real_type_2d_view fac("fac", 1, number_of_equations);
+
+  {
+      auto tol_time_host = Kokkos::create_mirror_view(tol_time);
+      auto tol_newton_host = Kokkos::create_mirror_view(tol_newton);
+      for (ordinal_type i = 0, iend = tol_time.extent(0); i < iend; ++i) {
+          tol_time_host(i, 0) = _atol_time;
+          tol_time_host(i, 1) = _rtol_time;
+      }
+      tol_newton_host(0) = _atol_newton;
+      tol_newton_host(1) = _rtol_newton;
+      
+      Kokkos::deep_copy(tol_time, tol_time_host);
+      Kokkos::deep_copy(tol_newton, tol_newton_host);
+  } 
+
+  using time_advance_type = TChem::time_advance_type;
+  time_advance_type tadv_default;
+  tadv_default._tbeg = 0.0;
+  tadv_default._tend = del_t;
+  tadv_default._dt = _dtmin;
+  tadv_default._dtmin = _dtmin;
+  tadv_default._dtmax = del_t;
+  tadv_default._max_num_newton_iterations = _max_num_newton_iterations;
+  tadv_default._num_time_iterations_per_interval = _num_time_iterations_per_interval;
+  tadv_default._jacobian_interval = _jacobian_interval;
+
+  time_advance_type_1d_view tadv("tadv", 1);
+  Kokkos::deep_copy(tadv, tadv_default);
+  real_type_1d_view t("time", 1);
+  Kokkos::deep_copy(t, tadv_default._tbeg);
+  real_type_1d_view dt("delta time", 1);
+  Kokkos::deep_copy(dt, tadv_default._dt);
+
+  real_type tend = del_t;
+  ordinal_type iter = 0;
+  real_type tsum(0);
+  auto stateVecDim = TChem_getLengthOfStateVector();
+  real_type_2d_view state("StateVector Devices", 1, stateVecDim);
+  Kokkos::deep_copy(state, _state);
+  for (; iter < _max_num_time_iterations && tsum <= tend * 0.9999; ++iter) {
+    TChem::AtmosphericChemistry::runDeviceBatch(policy, tol_newton, tol_time, fac, tadv,
+              state, t, dt, state, _kmcd_device);
+
+    tsum = 0.0;
+    Kokkos::parallel_reduce(
+        Kokkos::RangePolicy<TChem::exec_space>(0, 1),
+        KOKKOS_LAMBDA(const ordinal_type &i, real_type &update) {
+          tadv(i)._tbeg = t(i);
+          tadv(i)._dt = dt(i);
+          update += t(i);
+     }, tsum);
+     Kokkos::fence();
+  }
+  Kokkos::deep_copy(_state, state);
+}
+
+/* Return the state vector from host */
+void TChem::Driver::getStateVectorHost(real_type_2d_const_view_host &view) {
+  TCHEM_CHECK_ERROR(_state.span() == 0, "State vector should be constructed");
+  auto hv = Kokkos::create_mirror_view(_state);
```

Comment:
`_state` is a host view. I wonder if we need to create a mirror view.

Reply:
Fixed in 3c61a7c